### ISSUE TITLE
Never matches any type

### ISSUE THREAD

**Issue by bradzacher** [Microsoft/TypeScript#22683]
**TypeScript Version:**  2.7.2

**Search Terms:** "never" "never match"

**Code**

```ts
declare const x : never

function fn(v: string) { }

fn(x) // (1)
```

**Expected behavior:**
the call at `(1)` should fail with something along the lines of `never is not assignable to type string`.

**Actual behavior:**
the call is allowed

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=declare%20const%20x%20%3A%20never%0A%0Afunction%20fn(v%3A%20string)%20%7B%20%7D%0A%0Afn(x)%20%2F%2F%20(1)%0A

We ran into this because in `lib.dom.ts`, it defines a global variable called `name`, because of the global `window.name` property which stores the window's name if opened via `window.open`.

The problem is that one of our team members deleted the variable definition of `name` within a function. Typescript then fell back to using the global definition of name, which is `declare const name : never`.
[example of that exact case](https://www.typescriptlang.org/play/index.html#src=function%20fn(v%3A%20string)%20%7B%20%7D%0A%0Afn(name)%20%2F%2F%20(1)%0A)

Instead of the compiler throwing an error and disallowing `name` to be passed to a function, it instead let it pass.
This was caught in a code review, but in my opinion, it shouldn't have been able to get that far.

**Comment by RyanCavanaugh** [comment]
`never` is supposed to be assignable to any type.

See #18433 for an endless circular discussion about what the type of the global `name` should be

**Comment by bradzacher** [comment]
If you don't mind me asking - how come `never` is assignable to any type?

Imo that seems dangerous - there are examples where this can lead to unintended bugs in the code (i.e. the above code).

[This is an entirely convoluted example:](https://www.typescriptlang.org/play/index.html#src=function%20fn(x%3A%20string)%20%7B%0A%20%20%20%20if%20(typeof%20x%20%3D%3D%3D%20'string')%20%7B%0A%20%20%20%20%20%20%20%20throw%20'error'%0A%20%20%20%20%7D%0A%0A%20%20%20%20return%20x%0A%7D%0A%0Aconst%20ret%20%3A%20string%20%3D%20fn('foo')%0A%0A):
```TS
function fn(x: string) {
    if (typeof x === 'string') {
        throw 'error'
    }

    return x
}

const ret : string = fn('foo')
```
But realistically something like this could happen in the codebase, where a function returns `never` accidentally.

If that happens, the code will 100% compile fine, even though type-wise it makes zero sense.
The code that's passing around and manipulating a variable that, as far as types are concerned, shouldn't exist.

**Comment by RyanCavanaugh** [comment]
The code you've written is correct according to any type checker. `fn` is a function that always throws, so its return value cannot possibly assign any value into `ret`, so there is no possibility of an incorrect type being observed.

Any time you're assigning a value of type `never` to another value, then you're looking at code which *does not run* (according to the type system), so there is no problem.

See e.g. https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript for an example of how `never` is used in practice.

**Comment by RyanCavanaugh** [comment]
More specifically: it should always be possible to return a value of type `never` from any function regardless of what its return type is

**Comment by bradzacher** [comment]
I agree wholeheartedly that a function should be able to return `never`.

The question i have is this:
Why is `never` treated as an `any` during assignments/expressions?

```TS
const ret : string = fn('foo') // type never allowed to be assigned to type string 
const str = fn('foo') + '' // never + string = string
const val = fn('foo') + 1 // never + number = number
const boo = fn('foo') || false // never || boolean = boolean
```

You mention that it's okay because it does not run, but that relies on the types being 100% correct, which they're not always. (I.e. in the case of the `name` variable - the variable exists fine in the browser, but has been typed `never` in the browser lib typings.)

Also - isn't that all the more reason to have it throw a compiler error? Why would the default be to let me write code that has been detected to not run?

Sorry if I'm coming across in the wrong way; I'm just curious as to the reasoning behind it.

**Comment by RyanCavanaugh** [comment]
Consider something like this:
```ts
function callIfWorldIsEnding(x: () => number) { /* implementation unimportant */  }
callIfWorldIsEnding(() => {
  throw new Error("OH NO");
});
```
This code is clearly valid.

The function expression has type `never`. It's only a valid call to `callIfWorldIsEnding` if `never` is assignable to `number`. If `never` isn't assignable to `number`, then the call has to be rejected, which wouldn't really make any sense.

**Comment by RyanCavanaugh** [comment]
`never` is also *not* treated as `any` in the other direction, because you can't initialize a value of type `never` with any other type, *even `any`*:
```ts
// Error
const x: never = <any>30;
```
See the StackOverflow link above for why this has to be the case

**Comment by DanielRosenwasser** [comment]
It basically boils down to this:

1. For all union types *S*, and all types *T*, *S* is assignable to *T* if every constituent in *S* is assignable to *T*.
2. `never` is an empty union (i.e. a union with no constituents), so for all types *T*, it is [vacuously true](https://en.wikipedia.org/wiki/Vacuous_truth) that every constituent of `never` is assignable to *T*.
3. Therefore, for all types *T*, `never` is assignable to *T*.

**Comment by jack-williams** [comment]
@DanielRosenwasser You can similarly motivate `never` if you treat is as the intersection of all types also.
